### ISSUE TITLE
Intermittent 502 Bad Gateway w/ PHP 7.2+

### ISSUE THREAD

**Issue by BenceSzalai** [laravel/homestead#1654]
## Versions
  - Vagrant: 2.2.14
  - Provider: vagrant-parallels 2.2.0
  - Homestead: 11.0.0, 11.1.0, 11.3.0

## Host operating system
macOS 10.14.6

## Issue details
Right now I don't want to provide many details about my particular config, because this issue is only opened to see if anyone else is facing similar...? Normally I'd assume it is some kind of an issue with one of my kernel extensions or particular HW setup, because at the root `php-fpm` causes Segmentation faults but it is still rather strange that box `10.0.0` works just fine.

The issues appears as intermittent `502 Bad Gateway` error from the server, at approximately 60% of all requests.

Info:
- It does not depend on the PHP script executed. I am able to consistently reproduce it using a simple `<?php phpinfo();` as well.
- The same behaviour is expressed using nginx and apache (`type: "apache"`) as well.
- It is the same using box versions 11.0.0, 11.1.0, 11.3.0, however if I change `version` to `10.0.0` in `Homestead.yaml` there is no issue whatsoever with the exact same project.
- I haven't customised anything for the box, no custom scripts, installations anything. All is running on defaults and even after destroying the box and recreating from scratch the same happens.
- I have re-downloaded the base boxes, but it is the same.
- The issues is the same using PHP 7.2, 7.3, 7.4 and 8.0 as well, however **if I set PHP version to 5.6, 7.0 or 7.1 the system runs smooth and stable.**

Some info about the issue:
In `/var/log/nginx/myhostname.local-error.log` this line repeats:
```
2021/05/19 10:49:00 [error] 16638#16638: *1 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.10.1, server: myhostname.local, request: "GET info.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.4-fpm.sock:", host: "myhostname.local"
```

In php error log these repeat:
```
[19-May-2021 11:00:38] WARNING: [pool www] child 30735 exited on signal 11 (SIGSEGV - core dumped) after 11.263908 seconds from start
```

In the output of `dmesg` these appear:
```
[   89.757961] traps: php-fpm7.4[1208] general protection fault ip:7ff3ad7626ec sp:7ffea5196388 error:0 in libc-2.31.so[7ff3ad5f9000+178000]
[   94.470969] traps: php-fpm7.4[1209] general protection fault ip:7ff3ad7626ec sp:7ffea5196388 error:0 in libc-2.31.so[7ff3ad5f9000+178000]
[  178.927980] traps: php-fpm7.4[29914] general protection fault ip:7ff3ad7626ec sp:7ffea5196388 error:0 in libc-2.31.so[7ff3ad5f9000+178000]
[  791.198288] traps: php-fpm7.4[30735] general protection fault ip:7f698d9eb6ec sp:7ffc23880f98 error:0 in libc-2.31.so[7f698d882000+178000]
[ 2522.739233] traps: php-fpm7.4[32833] general protection fault ip:7f39a06566ec sp:7ffefefc4de8 error:0 in libc-2.31.so[7f39a04ed000+178000]
[ 3735.619523] traps: php-fpm7.4[34451] general protection fault ip:7f9b0bb766ec sp:7fffcc8eb6b8 error:0 in libc-2.31.so[7f9b0ba0d000+178000]
[ 4068.876084] traps: php-fpm7.4[34452] general protection fault ip:7f9b0ba80a2f sp:7fffcc8eba30 error:0 in libc-2.31.so[7f9b0ba0d000+178000]
[ 4073.051640] traps: php-fpm7.4[34464] general protection fault ip:7f9b0bb766ec sp:7fffcc8eb6b8 error:0 in libc-2.31.so[7f9b0ba0d000+178000]
[ 4167.418762] traps: php-fpm7.4[34801] general protection fault ip:7f9b0bb766ec sp:7fffcc8eb6b8 error:0 in libc-2.31.so[7f9b0ba0d000+178000]
[ 5369.499371] traps: php-fpm7.4[34804] general protection fault ip:7f9b0ba80a2f sp:7fffcc8eba30 error:0 in libc-2.31.so[7f9b0ba0d000+178000]
```

I have tried to install debug symbols as `sudo apt-get install php7.4-fpm-dbgsym` but it said it cannot be found, so I was not able to find out more about at where the fault is happening.

There is a similar issue reported here earlier: https://github.com/laravel/homestead/issues/1031

**Comment by BenceSzalai** [comment]
I have managed to get a backtrace:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:306
306     ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
(gdb) bt
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:306
#1  0x00007f8b3dc59257 in memcpy (__len=18, __src=0x7f8b3dc7da89, __dest=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:34
#2  xdebug_str_internal_addl (f=0, le=18, str=0x7f8b3dc7da89 "<tr><td class=\"i\">", xs=0x5607f784dbc0) at ./build-7.4/src/lib/str.c:46
#3  xdebug_str_addl (xs=0x5607f784dbc0, str=str@entry=0x7f8b3dc7da89 "<tr><td class=\"i\">", le=le@entry=18, f=f@entry=0) at ./build-7.4/src/lib/str.c:62
#4  0x00007f8b3dc57d98 in xdebug_diagnostic_log (
    message=0x7fff083cdf50 "The setting 'xdebug.remote_enable' has been renamed, see the upgrading guide at https://xdebug.org/docs/upgrade_guide#changed-xdebug.remote_enable", 
    error_code=0x7f8b3dc7b051 "CHANGED", log_level=0, channel=0) at ./build-7.4/src/lib/log.c:101
#5  xdebug_log_ex (channel=channel@entry=0, log_level=log_level@entry=0, error_code=error_code@entry=0x7f8b3dc7b051 "CHANGED", 
    fmt=fmt@entry=0x7f8b3dc7bb68 "The setting '%s' has been renamed, see the upgrading guide at %supgrade_guide#changed-%s") at ./build-7.4/src/lib/log.c:164
#6  0x00007f8b3dc503a4 in OnUpdateChangedSetting (new_value=<optimized out>, mh_arg1=<optimized out>, mh_arg2=<optimized out>, mh_arg3=<optimized out>, stage=<optimized out>, 
    entry=0x5607f7837b00) at ./build-7.4/xdebug.c:208
#7  OnUpdateChangedSetting (entry=0x5607f7837b00, new_value=<optimized out>, mh_arg1=<optimized out>, mh_arg2=<optimized out>, mh_arg3=<optimized out>, stage=<optimized out>)
    at ./build-7.4/xdebug.c:202
#8  0x00005607f5b61291 in zend_alter_ini_entry_ex ()
#9  0x00005607f5aedc7b in php_ini_activate_config ()
#10 0x00005607f5bdc837 in ?? ()
#11 0x00005607f5aeea6c in sapi_activate ()
#12 0x00005607f5ae556e in php_request_startup ()
#13 0x00005607f59ae9a9 in ?? ()
#14 0x00007f8b3fdd30b3 in __libc_start_main (main=0x5607f59adaf0, argc=4, argv=0x7fff083cebd8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fff083cebc8)
    at ../csu/libc-start.c:308
#15 0x00005607f59af81e in _start ()
```

So what it looks like to me is that PHP tries to enable xdebug, and xdebug has this outdated config key `xdebug.remote_enable` for which it genrates this "setting has been renamed" log entry, and during writing that log entry a `memcpy` writes to invalid address.

The question is how to find out if it is an issue with Homestead, Xdebug, PHP, Ubuntu or whatever?

**Comment by svpernova09** [comment]
Confirming you're seeing this on base box version 11.3.0, Parallels Provider?

Does #1652 seem familiar? But that would have been resolved via v11.3.0 base box.

**Comment by BenceSzalai** [comment]
This is what I see after ssh-ing in:
```
Welcome to Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-73-generic x86_64)
...
* Homestead v12.2.0 
* Settler v11.3.0 (Ubuntu 20.04)
```
I've checked the linked issue, but `php7.4-uopz` is not installed in `11.3.0`, so I don't really see a correlation.

I have the same issues with also 11.0.0 and 11.1.0 though.

**Comment by BenceSzalai** [comment]
I have updated Vagrant to v2.2.16 and Parallels provider to v2.2.1, but it didn't help.

**Comment by BenceSzalai** [comment]
Sorry for spamming the issue, but I've found the root cause and a way to avoid the issue, but I think there is still some strange underlying problems here.

I've got curious, why do I have this "setting has been renamed" errors from Xdebug in the first place???
My `/var/log/nginx/myhostname.local-error.log` was full of these:
```
2021/05/19 16:08:05 [error] 10355#10355: *6 FastCGI sent in stderr: "PHP message: Xdebug: [Config] The setting 'xdebug.remote_enable' has been renamed, see the upgrading guide at https://xdebug.org/docs/upgrade_guide#changed-xdebug.remote_enable (See: https://xdebug.org/docs/errors#CFG-C-CHANGED)PHP message: Xdebug: [Config] The setting 'xdebug.remote_autostart' has been renamed, see the upgrading guide at https://xdebug.org/docs/upgrade_guide#changed-xdebug.remote_autostart (See: https://xdebug.org/docs/errors#CFG-C-CHANGED)" while reading response header from upstream, client: 192.168.10.1, server: myhostname.local, request: "GET /info.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.4-fpm.sock:", host: "myhostname.local"
```

But when I look at the xdebug.ini files either in `/etc/php/7.4/xdebug.ini` or in `/etc/php/7.4/fpm/conf.d/20-xdebug.ini` I cannot see any Xdebug 2 settings, such as `xdebug.remote_enable` or `xdebug.remote_autostart`.

Then it hit me, that I have a very old `user.ini` in my project copied over from another project where it was copied as well etc. And surely inside that `user.ini` i had these:
```ini
[xdebug]
xdebug.remote_enable=true
xdebug.remote_autostart=true
```

Okay, so the good new is that if I remove these two lines, Xdebug no longer tries to write the "setting has been renamed" errors into the logs, and the segfault is gone, the system is stable now.

On the other hand though, I am asking a question, is it healthy that having an outdated ini setting, for which Xdebug is prepared in theory to react by logging deprecation message can cause the whole FPM process to crash? Probably not, and I feel like it is only the tip of an iceberg of something more nefarious deeper down the line.
